### PR TITLE
Update PJC to 2.11.2 for compatibility with new Panoptes updates

### PR DESCRIPTION
**Pull request by shaunanoordin**
## PR Overview
This PR updates `panoptes-client` to v2.11.2, which is necessary for compatibility with one of Panoptes's upcoming updates. See https://github.com/zooniverse/panoptes-javascript-client/pull/103 for further details.

### Status
❗️ This is a WIP as there's a huge jump from v0.1.2 to v2.11.2 - I'm looking into the changes necessary in `app/api/client.coffee`

### CODE REVIEW

**Comment by shaunanoordin** [comment]
## PR Update
@srallen This is as far as I've been able to get the PJC updates for WildCam Gorongosa, FYI.

I've ripped out the old `client = new PanoptesClient()` constructor code, and replaced them with direct imports to the client components, e.g. `oauth = require 'panoptes-client/lib/oauth'` as @eatyourgreens suggested.

The problem is, basically, Sign In still doesn't work - I _think_ I've properly managed to shunt users to the Panoptes sign in page (?) but users get stuck at panoptes.zooniverse.org.

Two things I've noted:
1. I'm able to run the Classifier just fine, provided I supply `?env=production`. (Simple, yes, but I spent half a day figuring out that there's no WCG on staging.) So at least parts of the API Clien't still ticking along as usual.
2. I can't seem to modify WildCam Gorongosa in Panoptes oAuth Apps (e.g. twiddling around with the accepted return URIs); any attempt to modify the app (e.g. to debug any "invalid redirect URI" errors) results in a 422 error. I believe Cam mentioned this is the result of a recent security update that doesn't... think localhost is kosher? I have to double check on this.

**Comment by srallen** [comment]
I remember now that how this is authenticating was before we even had auth in the panoptes-javascript-client. The files you see for setting up the client was literally copied and pasted from PFE at the time. It'll be best to just throw out what was done before and make updates to how it's done now. I started that process in the commit I made, but it's not done. ~However, I broke being able to set the env to production using url query param I think.~ I didn't break it; I just forgot that query params have to come before the hash.

**Comment by srallen** [comment]
Just to clarify, the code that WG was using that was copied and pasted from PFE was configured to use PFE's first party app id, etc. So that's part of the reason why all of that needs to be thrown out.

**Comment by shaunanoordin** [comment]
@srallen setting `env` via the query param works great, actually!

Observations
- When I access `http://localhost:3434/#/` with no `env` (i.e. default to staging), then I see the "Can't find workflow 338" as well as a butt tonne of other errors.
- When I access `http://localhost:3434/?env=production#/`, the console looks clear and, most importantly, I'm able to view the Classify page and submit a classification. 👍 
  - Just to check - did you type in `http://localhost:3434/#/?env=production` by accident? I kept doing that myself.

Observations Pt 2: Signing In Is Confusing
- Let's say I open a new Incognito window on Chrome.
- I access `http://localhost:3434/?env=production`, then click on the Sign In button.
- I'm taken to the sign in form on Panoptes (production, not staging 👌) and proceed to type in my code.
- I'm taken back to `http://localhost:3434/` (note the missing `env`) because that's the redirect URI, and as a result I'm in the wrong environment.
- I type in `http://localhost:3434/?env=production` in the URL and voila! I see that that the API Client is correctly recognising I'm trying to sign in via the console message `Got session darkeshard 1391493`
- However, the rest of the WCGorongosa website doesn't seem to recognise I'm supposed to be signed in, because the Sign In/Sign Up button is still there.

screen shot 2018-08-16 at 21 12 08
_What I see after I return from the Sign In page_

**Comment by shaunanoordin** [comment]
Wait, I just saw there's a new commit - my previous comments may be invalid now. brb

**Comment by shaunanoordin** [comment]
OK, I've ran tests and my previous comments still stand 👌 

One thing to note is that my `NODE_ENV` environmental variable is set to `development`, which may result in some non-standard testing experiences between computers. I thought this might be worth noting since I had funky experiences getting older projects like WCG to run before, because my NODE_ENV variable was blank or set to staging. (e.g. parts of some old code kept referring to `process.env.NODE_ENV` instead of looking at a global `env = getFromUrl() || process.env.NODE_ENV || DEFAULT_ENV` check.) 🤷‍♂️

**Comment by srallen** [comment]
I don't think it'll work for development until zooniverse/Panoptes#2901 or we could try to setup https for local dev.

**Comment by camallen** [comment]
Login should be working as is, https://github.com/zooniverse/Panoptes/pull/2901 is only for changing insecure redirect urls via the web forms. Happy to help debug this.

**Comment by shaunanoordin** [comment]
**TL;DR: looks like the back-end is working fine, but there's some front-end Reflux and CoffeeScript weirdness that I can't properly debug.**

I've ran tests on localhosts again, and this is what I'm seeing:

1. `Panoptes.zooniverse.org` (production) _is_ properly returning the User resource for the logged in user. (Tested on `localhost?env=production`)
2. However, some of the Reflux data updates don't seem to be firing properly.

**Weirdness 1: I can't tell why Coffeescript code isn't running**

```
//In app/stores/user-store.coffee...

71.  getUserClassificationCount: (user, _page = 1) ->
72.    query =
73.      user_id: user.id
74.      project_id: projectConfig.projectId
75.      page: _page
76.    api.type('project_preferences').get(query).then ([projectPreferences]) =>
77.      @createStore(user, projectPreferences)
```

I can tell that `getUserClassificationCount()` runs _and_ receives a User resource, but line `76` doesn't seem to be triggering - this can be tested either by checking to see if `@createStore` runs (it doesn't) or if `/project_preferences` is accessed in the network tab (it isn't).

Question: Why isn't line `76` running?

**Weirdness 2: Debugging Stuff**

Let's say I bypass line 76, and for testing purposes, I modify the code to look something like...

```
//In app/stores/user-store.coffee...

71.  getUserClassificationCount: (user, _page = 1) ->
72.    @createStore(user, null)
```

This sets the User resource/variable correctly (verified by how the app now starts to fetch the User's assignments), but the `main-header` component doesn't change to reflect this (i.e. the top header doesn't recognise that its `@props.user` has been updated, and it should show the User's profile instead of the Sign In button) 

Question: in `main-header.cjsx`, what's the Reflux equivalent to Redux's `connect()` method; i.e. what line of code tells the component to listen to changes in the `user-store` Store?

**Comment by srallen** [comment]
@camallen I was seeing CORS errors on Friday which made me think they were related. 

@shaunanoordin I'll take a look again. Reflux doesn't have an equivalent. The store is just assigned the component class itself, `@trigger` is the method on the store that should be updating the component classes.

**Comment by camallen** [comment]
> I was seeing CORS errors on Friday which made me think they were related.

Those CORS errors are semi-related and thanks for pointing them out. We don't allow localhost to authenticate to the production system. You can use local.zooniverse.org instead and remap that domain to point to your localhost server via DNS , e.g. `/etc/hosts`. 

I've updated https://github.com/zooniverse/Panoptes/pull/2901 to handle local.zooniverse.org http domains for testing against production systems.

**Comment by srallen** [comment]
@shaunanoordin login should now be working. There were several issues still remaining:

- The redirect url wasn't being set correctly depending on the environment
- Coffeescript has both fat arrow and skinny arrow functions. Fat vs skinny changes what `this` is in the method. The wrong type was being used in the sign out handler I refactored (hey it's been a while since I've written coffeescript)
- `api` wasn't being required, so it was undefined. I have no idea why this wasn't throwing a TypeError
- I added setting the node env on the npm start script. I'm not sure if this is really necessary, since we override this by using the query param environment setting to develop against production.

You should be able to see this all working using `http://local.zooniverse.org:3434/?env=production#/`

**Comment by shaunanoordin** [comment]
> `api` wasn't being required, so it was undefined. I have no idea why this wasn't throwing a TypeError

😱 why do you do this to us, Coffeescript???

**Comment by shaunanoordin** [comment]
## PR Review
Sweet! 👍 This PR rewrites several portions of the old CoffeeScript code (related to authentication) and brings WildCam Gorongosa up to date with the most recent version of the Panoptes JavaScript Client.
- Users can now login and logout via Panoptes oAuth
- WCG continues to properly respect the user's login state (e.g. logged-in user's account is shown on the top header, and Assignments are listed properly if the user has any.)

### Status
@srallen, a big ol' thank you for wrangling that CoffeeScript code into something that works. I'll be merging & deploying this shortly. 👌
